staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -768,6 +770,120 @@ func (r *Request) Watch(ctx context.Context) (watch.Interface, error) {
 	}
 }
 
+type WatchListResult struct {
+	// err holds any errors we might have received
+	// during streaming.
+	err error
+
+	// store holds collected data
+	store store
+
+	// initialEventsEndBookmarkRV holds the resource version
+	// extracted from the bookmark event that marks
+	// the end of the stream.
+	initialEventsEndBookmarkRV string
+}
+
+func (r WatchListResult) Into(obj runtime.Object) error {
+	if r.err != nil {
+		return r.err
+	}
+
+	listPtr, err := meta.GetItemsPtr(obj)
+	if err != nil {
+		return err
+	}
+	listVal, err := conversion.EnforcePtr(listPtr)
+	if err != nil {
+		return err
+	}
+	if listVal.Kind() != reflect.Slice {
+		return fmt.Errorf("need a pointer to slice, got %v", listVal.Kind())
+	}
+
+	listObjects := r.store.List()
+	if len(listObjects) == 0 {
+		listVal.Set(reflect.MakeSlice(listVal.Type(), 0, 0))
+	} else {
+		listVal.Set(reflect.MakeSlice(listVal.Type(), len(listObjects), len(listObjects)))
+		for i, o := range listObjects {
+			listVal.Index(i).Set(reflect.ValueOf(o).Elem())
+		}
+	}
+
+	listMeta, err := meta.ListAccessor(obj)
+	if err != nil {
+		return err
+	}
+	listMeta.SetResourceVersion(r.initialEventsEndBookmarkRV)
+
+	// TODO: set kind and apiversion
+	return nil
+}
+
+func (r *Request) WatchList(ctx context.Context) WatchListResult {
+	// assume the opts have been properly set
+	w, err := r.Watch(ctx)
+	if err != nil {
+		return WatchListResult{err: err}
+	}
+
+	return r.handleWatchList(ctx, w)
+}
+
+func (r *Request) handleWatchList(ctx context.Context, w watch.Interface) WatchListResult {
+	defer w.Stop()
+	s := newOrderedStore()
+	initialEventsEndBookmarkRV := ""
+
+	for {
```

Comment:
I still need to fill in details but it should be similar to what we have for the reflector - https://github.com/kubernetes/kubernetes/blob/b31e7793d0d873a71c90caf8455556aa905cf88d/staging/src/k8s.io/client-go/tools/cache/reflector.go#L655

Reply:
actually, it could be simplified, we should change the `cache` implementation to be consistent with the `etcd` implementation. the `etcd` currently sorts `range` request, the `cache` should follow suite.